sharktank/sharktank/layers/rotary_embedding_sharded.py

```diff
@@ -0,0 +1,156 @@
+            t_0 = ops.replicate(
+                t_0, self._tensor_parallelism_size, devices=self._devices
+            )
+            t_1 = ops.replicate(
+                t_1, self._tensor_parallelism_size, devices=self._devices
+            )
+
+        return t_0, t_1
+
+    def forward(
+        self,
+        *,
+        xt: Union[torch.Tensor, ShardedTensor],
+        start_index: int,
+    ):
+        t = torch.arange(xt.shape[1]).unsqueeze(0) + start_index
+        table_0, table_1 = self.rotary_embed_table(t)
+
+        if not isinstance(table_0, ShardedTensor):
+            return self._rotary_layer(q=xt, sincos_cache=(table_0, table_1))
+
+        shards = [
+            self._rotary_layer(q=xs, sincos_cache=(ts_0, ts_1))
+            for xs, ts_0, ts_1 in zip(xt.shards, table_0.shards, table_1.shards)
+        ]
+        return xt.clone(ts=shards)
+
+    def compute_batch_mask(
+        self, start_positions: Union[torch.Tensor, ShardedTensor], batch_seq_len: int
+    ) -> torch.Tensor:
+
+        if isinstance(start_positions, ShardedTensor):
+            shards_0 = []
+            shards_1 = []
+            for ss in start_positions.shards:
+                positions_seq = torch.arange(0, batch_seq_len)
+                positions_seq = positions_seq.unsqueeze(0) + ss.unsqueeze(1)
+                table_0, table_1 = self.rotary_embed_table(positions_seq)
+                shards_0.append(table_0)
+                shards_1.append(table_1)
+
+            return start_positions.clone(ts=shards_0), start_positions.clone(
+                ts=shards_1
+            )
+
+        positions_seq = torch.arange(0, batch_seq_len)
+        positions_seq = positions_seq.unsqueeze(0) + start_positions.unsqueeze(1)
+        table_0, table_1 = self.rotary_embed_table(positions_seq)
+        return table_0, table_1
+
+    def apply_batched_mask(
+        self,
+        *,
+        xt: Union[torch.Tensor, SplitPrimitiveTensor, ReplicatedTensor],
+        mask: tuple[Union[torch.Tensor, ReplicatedTensor]],
+    ) -> Union[SplitPrimitiveTensor, ReplicatedTensor]:
+
+        if not isinstance(xt, ShardedTensor):
+            return self._rotary_layer(q=xt, sincos_cache=mask)
+
+        assert (
+            isinstance(mask[0], ReplicatedTensor)
+            and mask[0].shard_count == xt.shard_count
+        )
+        assert (
+            isinstance(mask[1], ReplicatedTensor)
+            and mask[1].shard_count == xt.shard_count
+        )
+        xt_shards = [
+            self._rotary_layer(
+                q=unbox_tensor(xt_shard),
+                sincos_cache=(
+                    unbox_tensor(mask_sin_shard),
+                    unbox_tensor(mask_cos_shard),
+                ),
+            )
+            for xt_shard, mask_sin_shard, mask_cos_shard in zip(
+                xt.shards, mask[0].shards, mask[1].shards
+            )
+        ]
+        return xt.clone(ts=xt_shards)
```

Comment:
Let's not change the interface to rotary please. This interface is why i rewrote rotary. It's really bad. compute_batch_mask/apply_batch_mask/forward is bad. The only thing that needs to be there is compute_cache, and apply_rotary. nothing else please.